scanner.go

```diff
@@ -205,7 +207,21 @@ func fillInZeroedPtrs(ptrs []interface{}) {
 				elem.Set(reflect.MakeSlice(elem.Type(), 0, 0))
 			}
 		}
+	}
+}
 
+// removeSentinelValues removes any clustering sentinel values from being
+// exposed as data is scanned
+func removeSentinelValues(ptrs []interface{}) {
+	for _, ptr := range ptrs {
+		if _, ok := ptr.(*IgnoreFieldType); ok {
+			continue
+		}
+
+		elem := reflect.ValueOf(ptr).Elem()
+		if isSentinel, nonSentinelValue := isClusteringSentinelValue(elem.Interface()); isSentinel {
```

Comment:
to what degree (if any) should we be thinking about potential performance regressions by adding an extra step of reflection/value comparison for every field pointer now, do you think?

Reply:
We can benchmark this using the existing benchmarks

Before:
```
goos: darwin
goarch: amd64
pkg: github.com/monzo/gocassa
cpu: Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz
BenchmarkDecodeBlogSliceNoBody
BenchmarkDecodeBlogSliceNoBody-8       	   29456	     34640 ns/op
BenchmarkDecodeBlogStruct
BenchmarkDecodeBlogStruct-8            	  578539	      1841 ns/op
BenchmarkDecodeBlogStructEmptyBody
BenchmarkDecodeBlogStructEmptyBody-8   	  649267	      1840 ns/op
BenchmarkDecodeAlphaSlice
BenchmarkDecodeAlphaSlice-8            	   10000	    101197 ns/op
BenchmarkDecodeAlphaStruct
BenchmarkDecodeAlphaStruct-8           	  236494	      4842 ns/op
BenchmarkStatementMapTable
BenchmarkStatementMapTable-8           	  412954	      2773 ns/op	    3664 B/op	      33 allocs/op
PASS
ok  	github.com/monzo/gocassa	7.574s
```

After:
```
goos: darwin
goarch: amd64
pkg: github.com/monzo/gocassa
cpu: Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz
BenchmarkDecodeBlogSliceNoBody
BenchmarkDecodeBlogSliceNoBody-8       	   31040	     39350 ns/op
BenchmarkDecodeBlogStruct
BenchmarkDecodeBlogStruct-8            	  538323	      2429 ns/op
BenchmarkDecodeBlogStructEmptyBody
BenchmarkDecodeBlogStructEmptyBody-8   	  524773	      2916 ns/op
BenchmarkDecodeAlphaSlice
BenchmarkDecodeAlphaSlice-8            	    8732	    133401 ns/op
BenchmarkDecodeAlphaStruct
BenchmarkDecodeAlphaStruct-8           	  181543	      6072 ns/op
BenchmarkStatementMapTable
BenchmarkStatementMapTable-8           	  432660	      2886 ns/op	    3728 B/op	      33 allocs/op
PASS
ok  	github.com/monzo/gocassa	9.648s
```

I think we can probably optimise this codepath a bit more (the before is quite optimised since https://github.com/monzo/gocassa/pull/68)